lib/spack/spack/spec.py

```diff
@@ -2884,37 +2912,41 @@ def write(s, c):
                 named_str = named_str.upper()
                 if named_str == 'PACKAGE':
                     name = self.name if self.name else ''
-                    write(fmt % self.name, '@')
+                    write(fmt % tr[named_str](self.name), '@')
                 if named_str == 'VERSION':
                     if self.versions and self.versions != _any_version:
-                        write(fmt % str(self.versions), '@')
+                        write(fmt % tr[named_str](str(self.versions)), '@')
                 elif named_str == 'COMPILER':
                     if self.compiler:
-                        write(fmt % self.compiler, '%')
+                        write(fmt % tr[named_str](self.compiler), '%')
                 elif named_str == 'COMPILERNAME':
                     if self.compiler:
-                        write(fmt % self.compiler.name, '%')
+                        write(fmt % tr[named_str](self.compiler.name), '%')
                 elif named_str in ['COMPILERVER', 'COMPILERVERSION']:
                     if self.compiler:
-                        write(fmt % self.compiler.versions, '%')
+                        write(fmt % tr[named_str](self.compiler.versions), '%')
                 elif named_str == 'COMPILERFLAGS':
                     if self.compiler:
-                        write(fmt % str(self.compiler_flags), '%')
+                        write(
+                            fmt % tr[named_str](str(self.compiler_flags)), '%'
+                        )
                 elif named_str == 'OPTIONS':
-                    if self.variants:
-                        write(fmt % str(self.variants), '+')
+                    # Not all concretized specs have variants, but we still
+                    # want to run the transformation function on the variants
+                    if hasattr(self, 'variants'):
```

Comment:
If `bool(self.variants) == False` then `str(variants) == ""` - are you saying that the transformation function may still want to output something when given `""`? What makes this special vs. `Spec.compiler`?

Reply:
> Are you saying that the transformation function may still want to output something when given ""? 

Yes. The point for any hook using `Spec.format` is that in the context of a concretized spec, there's always a compiler but not always a set of variants (for instance `libszip` has no variants). Nonetheless, I still want to evaluate the custom transformation functions because they may have side effects - like raising if an invalid token is used in the format string. 

I think the code in `Spec.format` can probably benefit from a refactoring, but here I just tried to do the minimal changes needed to fix the two issues i.e.:

- make uppercase only the named tokens in the environment variable names
- report any use of invalid tokens

Reply:
> there's always a compiler but not always a set of variants (for instance libszip has no variants)

That is true in the implied context you are working with I suppose: at the time module files are generated, the package would be installed and would have a compiler. That being said I think it's a bit unclear what is going on here and that implied context could easily be lost later on, so I'd prefer a more direct approach. In other words I'd prefer it if future readers and editors of this code didn't have to think about the context where `Spec.format` was being called from.

> Nonetheless, I still want to evaluate the custom transformation functions because they may have side effects - like raising if an invalid token is used in the format string.

Are there other side effects you anticipate wanting? Generating an error is the only one I can think of. IMO it is better to make the effects explicit rather than depending on them to occur as side effects.

Suggestion: If the dictionary you pass only contained entries for tokens you are willing to write out, then this could detect a problem without having to invoke a function. I think that would be clearer and cover the only use case I'm currently aware of for spec properties that shouldn't be supported (for a particular invocation of `Spec.format`). You could also edit the method signature of `Spec.format` to optionally pass a customized error message, but actually IMO it would be more straightforward to throw an instance of `PropertyNotSupported` and catch it in the module file generator. 

Furthermore, regarding a request I made previously:

> I wouldn't require it of this PR, but it would be nice if it failed before calling functions like .use_name, like when the module writer was first initialized. Perhaps something simple like passing a non-concrete spec would work (for a package that doesn't actually exist).

If the dictionary only contained transform functions for supported Spec properties, then I think one could pass in an "empty" spec and detect issues with the format string before attempting to generate a module file name for a specific Spec. That's not necessarily motivating write now because the module file generators aren't instantiated until after a Package is installed, but if they were instantiated before package installations you could know there was an error with the format string before installing a package. To be clear: I'm not suggesting that the time when module file generators are instantiated should be changed here.

Reply:
> In other words I'd prefer it if future readers and editors of this code didn't have to think about the context where Spec.format was being called from.

That's actually a fair point. I restored the previous check in `Spec.format`, and scan for invalid tokens before calling it in the `spack.modules` subpackage. The solution is slightly different from what you propose but is, I think, along the same line.

> Are there other side effects you anticipate wanting? 

Maybe substitute a token with a default string if nothing is there? Right now a format like:
```python
'${PACKAGE}-${OPTIONS}-${VERSION}'
```
for e.g. `libszip@2.1.1` will return:
```python
'libszip--2.1.1'
```
and there's no way to get the `${OPTIONS}` part evaluated to a defautl, like:
```python
'libszip-None-2.1.1'
```
Anyhow, just thinking ahead for possible use cases: I don't need this functionality right now.